pkg/crd-schema/gen.go

```diff
@@ -0,0 +1,352 @@
+			}
+		} else {
+			changed, err = existing.setVersionedSchema(versionSchemas)
+			if err != nil {
+				return fmt.Errorf("failed to set versioned schemas for %s in %s: %v", groupKind.Kind, groupKind.Group, err)
+			}
+		}
+		if changed {
+			fmt.Printf("Updating schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		} else {
+			fmt.Printf("Unchanged schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		}
+
+		// check whether the file changes
+		if changed, err := existing.compareToExistingFile(fn); err != nil {
+			return fmt.Errorf("failed to compare against %q: %v", fn, err)
+		} else if changed {
+			changedFiles = append(changedFiles, fn)
+		}
+	}
+
+	// write YAML or check for differences
+	if g.VerifyOnly {
+		if len(changedFiles) > 0 {
+			return fmt.Errorf("unexpected changes to: %s", strings.Join(changedFiles, ", "))
+		}
+	} else {
+		for _, fn := range changedFiles {
+			fmt.Printf("Writing %s.\n", fn)
+			bs, err := yaml.Marshal(existingCRDsByFileNames[fn].Yaml)
+			if err != nil {
+				return err
+			}
+			if err := ioutil.WriteFile(fn, bs, 0644); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+type existingCRD struct {
+	GroupKind schema.GroupKind
+	Yaml      interface{}
+	CRD       *apiextensionsv1beta1.CustomResourceDefinition
+}
+
+func (e *existingCRD) compareToExistingFile(fn string) (bool, error) {
+	bs, err := ioutil.ReadFile(fn)
+	if err != nil {
+		return false, err
+	}
+	var old yaml.MapSlice
+	if err := yaml.Unmarshal(bs, &old); err != nil {
+		return false, err
+	}
+	y := e.Yaml.(yaml.MapSlice)
+	return !reflect.DeepEqual(old, y), nil
+}
+
+func (e *existingCRD) setGlobalSchema(schema *apiextensionsv1beta1.JSONSchemaProps) (bool, error) {
+	y, err := crdschemayaml.ToYAML(schema)
+	if err != nil {
+		return false, err
+	}
+	old, _, err := crdschemayaml.GetNestedFieldNoCopy(e.Yaml, "spec", "validation", "openAPIV3Schema")
+	if err != nil {
+		return false, err
+	}
+	if reflect.DeepEqual(old, y) {
+		return false, nil
+	}
+
+	if e.Yaml, err = crdschemayaml.SetNestedField(e.Yaml, y, "spec", "validation", "openAPIV3Schema"); err != nil {
+		return false, err
+	}
+	vs, found, err := crdschemayaml.GetNestedFieldSliceNoCopy(e.Yaml, "spec", "versions")
+	if err != nil {
+		return false, err
+	}
+	if found {
+		for i := range vs {
+			vs[i], err = crdschemayaml.DeleteNestedField(vs[i], "schema")
+			if err != nil {
+				return false, fmt.Errorf("spec.versions[%d].%s", i, err)
+			}
+		}
+		e.Yaml, err = crdschemayaml.SetNestedField(e.Yaml, vs, "spec", "versions")
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func (e *existingCRD) setVersionedSchema(versionSchemas map[string]*apiextensionsv1beta1.JSONSchemaProps) (bool, error) {
+	var err error
+	e.Yaml, err = crdschemayaml.DeleteNestedField(e.Yaml, "spec", "validation")
```

Comment:
why aren't we just deserializing into unstructured and using the equivalent methods there?

Reply:
actually, why aren't we deserializing into an `apiext.CustomResourceDefinition`?

Reply:
because it breaks order of maps and therefore destroys the existing files.